File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        cur = []

        def backtrack(open_count, close_count):
            if open_count == close_count == n:
                res.append(''.join(cur))
                return
            
            if open_count < n:
                cur.append('(')
                backtrack(open_count + 1, close_count)
                cur.pop()
            
            if close_count < open_count:
                cur.append(')')
                backtrack(open_count, close_count + 1)
                cur.pop()

        backtrack(0, 0)
        return res

File: 0022-generate-parentheses/0022-generate-parentheses.go
func generateParenthesis(n int) []string {

    var res []string
    
    cur := make([]byte, 0, 2*n)
    
    var backtrack func(openCount, closeCount int)

    backtrack = func (openCount, closeCount int) {
        if openCount == n && closeCount == n {
            res = append(res, string(cur))
            return
        }

        if openCount < n {
            cur = append(cur, '(')
            backtrack(openCount + 1, closeCount)
            cur = cur[:len(cur)- 1]
        }

        if closeCount < openCount {
            cur = append(cur, ')')
            backtrack(openCount, closeCount + 1)
            cur = cur[:len(cur)- 1]
        }
    }

    backtrack(0, 0)

    return res
}